packages/wonder-blocks-button/components/clickable-behavior.js

```diff
@@ -0,0 +1,190 @@
+ *                      {this.props.children}
+ *                 </RoundRect>
+ *             }
+ *         </ClickableBehavior>
+ *     }
+ * }
+ *
+ * This follows a pattern call 'Function as Child Components':
+ * https://medium.com/merrickchristensen/function-as-child-components-5f3920a9ace9
+ */
+import React from "react";
+
+type Props = {
+    children: (state: State, handlers: Handlers) => React$Element<*>,
+    disabled: boolean,
+    href?: string,
+    onClick?: (e: SyntheticMouseEvent<>) => void,
+};
+
+type State = {
+    hovered: boolean,
+    focused: boolean,
+    pressed: boolean,
+};
+
+type Handlers = {
+    onClick: (e: SyntheticMouseEvent<>) => void,
+    onMouseEnter: () => void,
+    onMouseLeave: () => void,
+    onMouseDown: () => void,
+    onMouseUp: () => void,
+    onTouchStart: () => void,
+    onTouchEnd: () => void,
+    onKeyDown: (e: KeyboardEvent) => void,
+    onKeyUp: (e: KeyboardEvent) => void,
+    onBlur: () => void,
+};
+
+const keyCodes = {
+    tab: 9,
+    enter: 13,
+    space: 32,
+};
+
+export default class ClickableBehavior extends React.Component<Props, State> {
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            hovered: false,
+            focused: false,
+            pressed: false,
+        };
+    }
+
+    waitingForClick: boolean;
+
+    handleClick = (e: SyntheticMouseEvent<>) => {
+        if (!this.props.disabled && this.props.onClick) {
+            this.waitingForClick = false;
+            this.props.onClick(e);
+        }
+    };
+
+    handleMouseEnter = () => {
+        if (!this.props.disabled && !this.waitingForClick) {
+            this.setState({hovered: true});
+        }
+    };
+
+    handleMouseLeave = () => {
+        if (!this.props.disabled && !this.waitingForClick) {
+            // TODO(yejia): Is this what we want for pressed?
+            this.setState({hovered: false, pressed: false});
+        }
+    };
+
+    handleMouseDown = () => {
+        if (!this.props.disabled) {
+            this.setState({pressed: true});
+        }
+    };
+
+    handleMouseUp = () => {
+        if (!this.props.disabled) {
+            this.setState({pressed: false});
+        }
+    };
+
+    handleTouchStart = () => {
+        if (!this.props.disabled) {
+            this.setState({pressed: true});
+        }
+    };
+
+    handleTouchEnd = () => {
+        if (!this.props.disabled) {
+            this.setState({pressed: false});
+            this.waitingForClick = true;
+        }
+    };
+
+    handleKeyDown = (event: KeyboardEvent) => {
+        const keyCode = event.which || event.keyCode;
+        const {disabled, href} = this.props;
+        if (!disabled) {
+            if (keyCode === keyCodes.tab) {
+                this.setState({focused: false});
+            } else if (
+                href ? keyCode === keyCodes.enter : keyCode === keyCodes.space
+            ) {
+                this.setState({pressed: true});
```

Comment:
This worked for the first button but not for the second button for me when I was testing in Chrome.

Reply:
BTW, this is really cool!

Reply:
I think what's happening for the second one is that pressing enter immediately clicks the link, which takes you to the section linked and takes you out of focus of the button, so you're no longer "pressing" within the button. I think this is standard behavior for a link, but should the "click" happen onKeyUp?